src/App.tsx

```diff
@@ -203,6 +300,86 @@ class Board extends React.Component<BoardProps, BoardState> {
+            values: currentState.values,
+        };
+        console.log("new: " + Array.from(instanceToAdd.values.entries()));
+        if (_.isEqual(instanceToAdd.values, oldInstance.values) &&
+            _.isEqual(instanceToAdd.markingMap, oldInstance.markingMap)) {
+            return newState;
+        }
+        currentState.instance.nextInstance = instanceToAdd;
+        newState.instance = instanceToAdd;
+        /*newState.historyIndex = currentState.historyIndex + 1;
+        // On any history update the future history must be
+        // deleted because it is no longer the future of the
+        // new current (watch Back to the Future)
+        newState.history.splice(newState.historyIndex, currentState.history.length);
+        newState.history.push(instanceToAdd);*/
+        return newState;
+    }
+
+    public undo(): void {
+        const newState = _.cloneDeep(this.state);
+        if (newState.instance.previousInstance === null) {
+            return;
+        }
+        newState.instance = newState.instance.previousInstance;
+        newState.markingMap = _.cloneDeep(newState.instance.markingMap);
+        newState.values = _.cloneDeep(newState.instance.values);
+        this.setState(newState);
+        /*newState.historyIndex -= 1;
+        newState.markingMap = _.cloneDeep(newState.history[newState.historyIndex].markingMap);
+        newState.values = _.cloneDeep(newState.history[newState.historyIndex].values);*/
+    }
+
+    public redo(): void {
+        const newState = _.cloneDeep(this.state);
+        if (newState.instance.nextInstance) {
+            newState.instance = newState.instance.nextInstance;
+            newState.markingMap = _.cloneDeep(newState.instance.markingMap);
+            newState.values = _.cloneDeep(newState.instance.values);
+        }
+        this.setState(newState);
+        /*if (this.state.historyIndex + 1 === this.state.history.length) {
+            return;
+        }
+        const newState = _.cloneDeep(this.state);
+        newState.historyIndex += 1;
+        newState.markingMap = _.cloneDeep(newState.history[newState.historyIndex].markingMap);
+        newState.values = _.cloneDeep(newState.history[newState.historyIndex].values);
+        */
+    }
+
+    public restart(): void {
+        const reset = window.confirm("Are you sure you would like to restart? All saved progress will be lost.");
+        if (reset) {
+            const newState: BoardState = {
+                contradicts: new Set(),
+                highlights: new Set(),
+                instance: {
+                    markingMap: new Map(),
+                    nextInstance: null,
+                    previousInstance: null,
+                    values: new Map(),
+                },
+                markingMap: new Map(),
+                mouseOverHighlighting: false,
+                multiStrokeHighlighting: false,
+                numpadMode: "normal",
+                values: new Map(),
```

Comment:
Let's create a constant called `initialBoardState` that you can also use in the `BoardState` constructor.